### PR TITLE
Extend ROUND() and FLOOR() to include number of decimal places to round up/down #2583

### PR DESCRIPTION
**Pull request by ouertani**
### Description 
Fixes https://github.com/confluentinc/ksql/issues/2583 

### Testing done 
Test case added  https://github.com/confluentinc/ksql/commit/bbde8e8059b5ab67ef84d3e205fe118c4cb28259#diff-a6c1cc751c41072e368e476def856e6b

### Reviewer checklist
- [X] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
https://github.com/confluentinc/ksql/commit/79970a06113f7a0e51b6705b219d0ad58e689891
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: docs/developer-guide/syntax-reference.rst [hunk 1]
```diff
@@ -1512,7 +1512,11 @@ Scalar functions
 | ARRAYCONTAINS          |  ``ARRAYCONTAINS('[1, 2, 3]', 3)``                                        | Given JSON or AVRO array checks if a search       |
 |                        |                                                                           | value contains in it                              |
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
-| CEIL                   |  ``CEIL(col1)``                                                           | The ceiling of a value.                           |
+| CEIL                   |  ``CEIL(col1)`` or   ``CEIL(col1, precision)``                            | The ceiling of a value.                           |
+|                        |                                                                           |                                                   |
+|                        |                                                                           | precision is a number indicating the number of    |
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: "...number indicating the number..." maybe we can reword this?

How about: "precision is an INT indicating the number of decimal places to round up to."

File: docs/developer-guide/syntax-reference.rst [hunk 2]
```diff
@@ -1561,7 +1565,11 @@ Scalar functions
 |                        |                                                                           | considered to be equal to any value. FIELD is the |
 |                        |                                                                           | complement to ELT.                                |
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
-| FLOOR                  |  ``FLOOR(col1)``                                                          | The floor of a value.                             |
+| FLOOR                  |  ``FLOOR(col1)`` or   ``FLOOR(col1, precision)``                          | The floor of a value.                             |
+|                        |                                                                           |                                                   |
+|                        |                                                                           | precision is a number indicating the number of    |
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: How about: "precision is an INT indicating the number of decimal places to round down to."

File: docs/developer-guide/syntax-reference.rst [hunk 3]
```diff
@@ -1618,7 +1626,10 @@ Scalar functions
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
 | RANDOM                 |  ``RANDOM()``                                                             | Return a random DOUBLE value between 0.0 and 1.0. |
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
-| ROUND                  |  ``ROUND(col1)``                                                          | Round a value to the nearest BIGINT value.        |
+| ROUND                  |  ``ROUND(col1)`` or                                                       | Round a value to the nearest BIGINT value.        |
+|                        |  ``ROUND(col1, precision)``                                               |                                                   |
+|                        |                                                                           | precision is a number indicating up to how many   |
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: How about: "precision is an INT indicating the number of decimal places to round to."

(For consistency)

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 4]
```diff
@@ -233,21 +233,39 @@ private void addMathFunctions() {
           AbsKudf.class));
 
       addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
-          Schema.OPTIONAL_FLOAT64_SCHEMA,
+          Schema.OPTIONAL_INT64_SCHEMA,
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: I think this change would break backwards compatibility...

If customers already have a query running that uses the old code, when then upgrade the server will restart their query using the new code, but that will change the type of the result, which may well be incompatible.

We don't really have a way to version such UDFs to allow the return value to change between versions yet.

The _safe_ option is probably to stick with the old _incorrect_ return type for now. (Unless anyone has any ideas?) And just raise a Github issue to track, mentioning that just changing the return type would be a breaking change, so what we need first is a way to version UDFs and their result schemas.

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 5]
```diff
@@ -233,21 +233,39 @@ private void addMathFunctions() {
...
           CeilKudf.class));
 
       addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
           Schema.OPTIONAL_FLOAT64_SCHEMA,
+          ImmutableList.of(Schema.FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
+          CeilKudf.NAME,
+          CeilKudf.class));
+
+      addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
+          Schema.OPTIONAL_INT64_SCHEMA,
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: likewise, I think this is a breaking change.

File: ksql-functional-tests/src/test/resources/query-validation-tests/round-floor.json [hunk 6]
```diff
@@ -0,0 +1,82 @@
+{
+  "comments": [
+    "Tests covering the use of the ROUND(DOUBLE), ROUND(DOUBLE,INT), FLOOR(DOUBLE), FLOOR(DOUBLE,INT) function."
+  ],
+  "tests": [
+    {
+      "name": "calculate ROUND function",
+      "statements": [
+        "CREATE STREAM test (v DOUBLE) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT ROUND(v) as R0, ROUND(v, 0) as R00, ROUND(v, 1) as R1, ROUND(v, 2) as R2, ROUND(v, 10) as R10 FROM test;"
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/round-floor.json]: some impls of `ROUND` support negative numbers of the second argument, e.g.

`ROUND(124, -1)` would give `120`.
`ROUND(124, -2)` would give `100`.

Might be worth adding support for this.

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 7]
```diff
@@ -233,21 +233,39 @@ private void addMathFunctions() {
...
-          Schema.OPTIONAL_FLOAT64_SCHEMA,
+          Schema.OPTIONAL_INT64_SCHEMA,
           Collections.singletonList(Schema.OPTIONAL_FLOAT64_SCHEMA),
-          "CEIL",
+          CeilKudf.NAME,
           CeilKudf.class));
 
       addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
           Schema.OPTIONAL_FLOAT64_SCHEMA,
+          ImmutableList.of(Schema.FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]:
```suggestion
          ImmutableList.of(Schema.OPTIONAL_FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
```

We should support a `null` value here too.

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 8]
```diff
@@ -233,21 +233,39 @@ private void addMathFunctions() {
...
+      addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
+          Schema.OPTIONAL_INT64_SCHEMA,
           Collections.singletonList(Schema.OPTIONAL_FLOAT64_SCHEMA),
-          "FLOOR",
+          FloorKudf.NAME,
+          FloorKudf.class));
+
+      addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
+          Schema.OPTIONAL_FLOAT64_SCHEMA,
+          ImmutableList.of(Schema.FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: Likewise,

```suggestion
          ImmutableList.of(Schema.OPTIONAL_FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
```

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java [hunk 9]
```diff
@@ -233,21 +233,39 @@ private void addMathFunctions() {
...
       addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
           Schema.OPTIONAL_INT64_SCHEMA,
           Collections.singletonList(Schema.OPTIONAL_FLOAT64_SCHEMA),
-          "ROUND",
+          RoundKudf.NAME,
+          RoundKudf.class));
+
+      addBuiltInFunction(KsqlFunction.createLegacyBuiltIn(
+          Schema.OPTIONAL_FLOAT64_SCHEMA,
+          ImmutableList.of(Schema.FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: likewise,

```suggestion
          ImmutableList.of(Schema.OPTIONAL_FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
```

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/CeilKudf.java [hunk 10]
```diff
@@ -20,11 +20,20 @@
...
-    if (args.length != 1) {
-      throw new KsqlFunctionException("Ceil udf should have one input argument.");
+    if (args.length != 1 && args.length != 2) {
+      throw new KsqlFunctionException("Ceil udf should have one or two input arguments.");
+    }
+
+    final Double number = (Double) args[0];
+
+    if (args.length == 1) {
+      return (long) Math.ceil(number);
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/CeilKudf.java]: will result in NPE if `number` is null.   This would result in a log of errors being logged unnecessarily. We should just return `null` if number is `null`.

A unit test would be good to test this case, as QTT tests can't catch such an issue.  

We should also check the second arg for `null`.  If provided, but `null`... then what we do is debatable. Maybe just ignore it, i.e. a `null` second arg just the same as having only 1 arg.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/FloorKudf.java [hunk 11]
```diff
@@ -19,12 +19,18 @@
...
   public Object evaluate(final Object... args) {
-    if (args.length != 1) {
-      throw new KsqlFunctionException("Floor udf should have one input argument.");
+    if (args.length != 1  && args.length != 2) {
+      throw new KsqlFunctionException("Floor udf should have one or two input arguments.");
     }
-    return Math.floor((Double) args[0]);
+    final Double number = (Double) args[0];
+    if (args.length == 1) {
+      return (long) Math.floor(number);
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/FloorKudf.java]: will result in NPE if `number` is null.   This would result in a log of errors being logged unnecessarily. We should just return `null` if number is `null`.

A unit test would be good to test this case, as QTT tests can't catch such an issue.  

We should also check the second arg for `null`.  If provided, but `null`... then what we do is debatable. Maybe just ignore it, i.e. a `null` second arg just the same as having only 1 arg.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java [hunk 12]
```diff
@@ -19,12 +19,20 @@
...
   @Override
   public Object evaluate(final Object... args) {
-    if (args.length != 1) {
-      throw new KsqlFunctionException("Len udf should have one input argument.");
+    if (args.length != 1 && args.length != 2) {
+      throw new KsqlFunctionException("Round udf should have one or two input arguments.");
     }
-    return Math.round((Double) args[0]);
+
+    final Double number = (Double) args[0];
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: will result in NPE if `number` is null.   This would result in a log of errors being logged unnecessarily. We should just return `null` if number is `null`.

A unit test would be good to test this case, as QTT tests can't catch such an issue.  

We should also check the second arg for `null`.  If provided, but `null`... then what we do is debatable. Maybe just ignore it, i.e. a `null` second arg just the same as having only 1 arg.

File: docs/developer-guide/syntax-reference.rst (hunk 1 above)
Comment (spena) [reply docs/developer-guide/syntax-reference.rst]: Isn't `precision` confusing with the new DECIMAL type precision? DECIMAL has (precision, scale), and scale is the number of digits after the decimal point. Should we use a similar parameter name?

File: docs/developer-guide/syntax-reference.rst (hunk 2 above)
Comment (spena) [reply docs/developer-guide/syntax-reference.rst]: Same precision name opinion as `CEIL` (see above)

File: docs/developer-guide/syntax-reference.rst (hunk 3 above)
Comment (spena) [reply docs/developer-guide/syntax-reference.rst]: Same precision name opinion as CEIL (see above)

File: docs/developer-guide/syntax-reference.rst (hunk 1 above)
Comment (big-andy-coates) [reply docs/developer-guide/syntax-reference.rst]: scale works for me

**Comment by big-andy-coates** [comment]
Hey @spena 

> These UDF were useful for Decimals types, but I think the new DECIMAL implementation will have a round type as a 3rd parameter, which makes these UDF not necessary anymore. However, these might be useful for other cases.

Not sure I follow what you mean, do you mean `DECIMAL` implementation of `floor` etc?

@agavra, given you're working on DECIMAL support - what are your thoughts?

**Comment by blueedgenick** [comment]
Having a _scale_ factor for `CEIL()` and `FLOOR()` is very unusual - i just checked postgres/mysql/oracle/db2slqserver and NONE of them do. For consistency of expecations, can we confine this only to the `ROUND()` function please ?

**Comment by agavra** [comment]
I think I agree with @blueedgenick and @big-andy-coates that we should limit the scale factor to just round. `Truncate` and `Round` usually take in scale parameters, and you can achieve nearly anything using those two as you can with `Ceil` and `Floor`.

Unfortunately, even as it stands this change is backwards incompatible even just for `ROUND`. Previously, `ROUND` returned a `long` or `int` and now it will potentially return a `double` - any KSQL application that is running will not be able to upgrade since the schema will be backwards compatible.

The only way I see to do this would be to move to the new UDF framework, and adding a strongly typed method for each input combo:
```
public long round(float val);
public double round(float val, int scale);
public long round(double val);
public double round(double val, int scale);
...
```

**Comment by hjafarpour** [comment]
I also agree with keeping consistent with common practice as @blueedgenick mentioned. Let's male this change only for `ROUND()`.

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java (hunk 4 above)
Comment (ouertani) [reply ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: Yes, I was wondering  to correct it or not. The issue https://github.com/confluentinc/ksql/issues/2583 highlights the fact that FLOOR(3.6456) = 3 so not 3.0. I would recommend  to tackle it in separate thread.

**Comment by ouertani** [comment]
Last commit reverts back modifications on `FLOOR and CEIL` only ROUND has been considered.

File: ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java (hunk 9 above)
Comment (big-andy-coates) [reply ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java]: As mentioned before...

```suggestion
          ImmutableList.of(Schema.OPTIONAL_FLOAT64_SCHEMA, Schema.OPTIONAL_INT32_SCHEMA),
```

File: ksql-functional-tests/src/test/resources/query-validation-tests/array-functions.json [hunk 13]
```diff
@@ -0,0 +1,82 @@
+{
+  "comments": [
+    "Tests covering the use of the ROUND(DOUBLE), FLOOR(DOUBLE), CEIL(DOUBLE) functions."
+  ],
+  "tests": [
+    {
+      "name": "calculate ROUND function",
+      "statements": [
+        "CREATE STREAM test (v DOUBLE) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT ROUND(v) as R0, ROUND(v, 0) as R00, ROUND(v, 1) as R1, ROUND(v, 2) as R2, ROUND(v, 10) as R10 FROM test;"
```
Comment (big-andy-coates) [inline ksql-functional-tests/src/test/resources/query-validation-tests/array-functions.json]: The implementation you've added supports negative scales e.g. `ROUND(111, -1)` -> `110`.

Can we add a test case / column for this please?

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java [hunk 14]
```diff
@@ -20,11 +20,27 @@
...
+    if (args.length != 1 && args.length != 2) {
+      throw new KsqlFunctionException("Round udf should have one or two input arguments.");
+    }
+
+    if (args[0] == null) {
+      return null;
+    }
+    final Double number = (Double) args[0];
+
+    if (args.length == 1) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: As mentioned previously, I think we should treat the second parameter being `null` the same as there being only 1 parameter, so...

```suggestion
    if (args.length == 1 || args[1] == null) {
```

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java [hunk 15]
```diff
@@ -20,11 +20,27 @@
...
+      return Math.round(number);
+    }
+
+    if (args[1] == null) {
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: As above, remove this `if`

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java [hunk 16]
```diff
@@ -20,11 +20,27 @@
 
 public class RoundKudf implements Kudf {
 
+  public static final String NAME = "ROUND";
```
Comment (big-andy-coates) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: This class would benefit from a unit test. The JSON test is a functional test to ensure everything is wired up correctly, but a unit test is still a good thing and can test things the functional test can't, e.g. that it throws on the wrong params count / type.

File: docs/developer-guide/syntax-reference.rst [hunk 17]
```diff
@@ -1620,7 +1620,9 @@ Scalar functions
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
 | RANDOM                 |  ``RANDOM()``                                                             | Return a random DOUBLE value between 0.0 and 1.0. |
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
-| ROUND                  |  ``ROUND(col1)``                                                          | Round a value to the nearest BIGINT value.        |
+| ROUND                  |  ``ROUND(col1)`` or ``ROUND(col1, precision)``                            | Round a value to the nearest BIGINT value.        |
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
| ROUND                  |  ``ROUND(col1)`` or ``ROUND(col1, scale)``                            | Round a value to the nearest BIGINT value.        |
```

File: docs/developer-guide/syntax-reference.rst [hunk 18]
```diff
@@ -1620,7 +1620,9 @@ Scalar functions
...
+|                        |                                                                           | precision is an INT indicating the number.        |
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]:
```suggestion
|                        |                                                                           | scale is an INT indicating the number.        |
```

Comment (big-andy-coates) [reply docs/developer-guide/syntax-reference.rst]: I'd also document that a negative number will round to the left of the decimal point, e.g. 

`ROUND(111, -1)` results in `110`.

File: ksql-functional-tests/src/test/resources/query-validation-tests/array-functions.json (hunk 13 above)
Comment (big-andy-coates) [reply ksql-functional-tests/src/test/resources/query-validation-tests/array-functions.json]: Can you also add a test case/column where the second param to `ROUND` is `null`.  You can do this explicitly in the statement, or have the second param also come from an input field:

```
CREATE STREAM test (v DOUBLE, scale INT) WITH (kafka_topic='test_topic', value_format='JSON');
CREATE STREAM OUTPUT AS SELECT ROUND(v, scale) FROM test;
```

Then have inputs:
```
"inputs": [
        {"topic": "test_topic", "value": {"v" : 1.2, "scale": 0}},
        {"topic": "test_topic", "value": {"v" : 1.7, "scale": -1}},
        {"topic": "test_topic", "value": {"v" : 1.7, "scale": null}},
```

File: docs/developer-guide/syntax-reference.rst (hunk 18 above)
Comment (ouertani) [reply docs/developer-guide/syntax-reference.rst]: https://github.com/confluentinc/ksql/pull/3028/commits/ad0b1dd2a17a76771986263b2c3f83a35d853e1c

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java (hunk 16 above)
Comment (ouertani) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: https://github.com/confluentinc/ksql/pull/3028/commits/d04b758a72b748f35a5185b6ca8e97da61daf15e

File: docs/developer-guide/query-with-arrays-and-maps.rst [hunk 19]
```diff
@@ -104,6 +104,35 @@ Your output should resemble:
...
+
+.. code:: sql
+
+      SELECT interests[-1] AS last_interest,
+             userid,
+             gender,
+             regionid
+      FROM users_extended;
+
+Your output should resemble:
```
Comment (big-andy-coates) [inline docs/developer-guide/query-with-arrays-and-maps.rst]: Looks like you're merge has gone wonky...

File: docs/developer-guide/syntax-reference.rst [hunk 20]
```diff
@@ -1620,7 +1620,9 @@ Scalar functions
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
 | RANDOM                 |  ``RANDOM()``                                                             | Return a random DOUBLE value between 0.0 and 1.0. |
 +------------------------+---------------------------------------------------------------------------+---------------------------------------------------+
-| ROUND                  |  ``ROUND(col1)``                                                          | Round a value to the nearest BIGINT value.        |
+| ROUND                  |  ``ROUND(col1)`` or ``ROUND(col1, scale)``                            | Round a value to the nearest BIGINT value.        |
```
Comment (big-andy-coates) [inline docs/developer-guide/syntax-reference.rst]: nit: formating of the columns is off.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/math/RoundKudfTest.java [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    new KudfTester(AbsKudf::new)
+        .withArgumentTypes(Number.class)
+        .test();
+  }
+
+  @Test
+  public void shouldReturnNullWhenArgNull() {
+    assertThat(udf.evaluate((Object)null), is(nullValue()));
+  }
+
```
Comment (big-andy-coates) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/math/RoundKudfTest.java]: add test cases for:
-  second arg being null.
- no args
- too many args

**Comment by purplefox** [comment]
I'm going to attempt to get this working with current master

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java [hunk 22]
```diff
@@ -20,11 +20,25 @@
...
     }
-    return Math.round((Double) args[0]);
+    final Object value = args[0];
+    final Double number = (Double) value;
+
+    if (args.length == 1 || args[1] == null) {
+      return Math.round(number);
+    }
+
+    final Double round = Math.pow(10, (Integer) args[1]);
```
Comment (purplefox) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: Unfortunately this won't work in all cases due the way floating point numbers (e.g. double) represent numbers.
Consider the following:
```
double mult = 265.335d * 100d;
System.out.println(mult);
double res = Math.round(mult) / 100d;
System.out.println(res);
```

This outputs:
```
26533.499999999996
265.33
```

I.e it rounds down not up as 26533.5 cannot be represented exactly as a double.

Instead a DecimalFormat or BigDecimal can be used to give correct rounding.

Please see the discussion here for more details https://stackoverflow.com/questions/153724/how-to-round-a-number-to-n-decimal-places-in-java

Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: I think this can also be done by accounting for the error in the double. Most likely this would be faster than using BigDecimal or DecimalFormat. Something like: 
```
public static double round(double num, int decimalPlaces) {

    int pow10 = 1;
    for (int i = 0; i < decimalPlaces; i++) {
      pow10 *= 10;
    }

    double dPow10 = pow10;

    double top = num * dPow10;

    double ulp = Math.ulp(top);
    
    return  Math.round(top + ulp) / dPow10;
    
  }
```
Would need testing, esp. with edge cases, e.g. probably doesn't work with -ve num and -ve decimal places as is.

Comment (purplefox) [reply ksql-engine/src/main/java/io/confluent/ksql/function/udf/math/RoundKudf.java]: Having said that, probably the safest thing to do is to use BigDecimal :)

**Comment by purplefox** [comment]
Closing this as it has been superseded by https://github.com/confluentinc/ksql/pull/3404

**Comment by big-andy-coates** [comment]
@ouertani just wanted to call out a big "thank you" for this PR.  @purplefox has finished it off for you and its now merged!
